x/programs/program/reentrancy.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package program
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+)
+
+var ErrReentrancy = errors.New("reentrant call not allowed")
+
+type reentrancyMap map[ids.ID]map[string]bool
```

Comment:
🤔, on a re-entrant call, do instantiate a new call context? 

If we have program A that calls program B that calls back into A, do we instantiate two As? 

If so, we probably shouldn't do that? 

I think this might help with the whole context thing. We shouldn't really need a map of maps, we should only need a map attached to the call context and/or program or whatever. 

Do I know how to do that? Absolutely not haha. It's definitely possible though. 

Also curious about @dboehm-avalabs's thoughts here.

Reply:
yea, I agree it should be attached on a per context level. However, I'm not sure how to attach it... Wasmtime requires this function signature for imported functions
`fn := func(caller *wasmtime.Caller, args []wasmtime.Val) ([]wasmtime.Val, *wasmtime.Trap) {`

ideally we could pass along our reentrancy data into `caller`

Reply:
I case it's not clear from the fact that I approved this, I'm not suggesting you make any of the changes from this approval now. I left another comment somewhere saying that we could actually do ALL of the locking, etc, in the guest, and there would be fewer host calls. 

Of course, that's contingent on us not spinning multiple instances of the same program...